src/frontend/src/optimizer/plan_node/logical_changelog.rs

```diff
@@ -66,15 +66,26 @@ impl PlanTreeNodeUnary for LogicalChangeLog {
         input_col_change: ColIndexMapping,
     ) -> (Self, ColIndexMapping) {
         let changelog = Self::new(input, self.core.need_op, true);
-        if self.core.need_op {
+
+        let out_col_change = if self.core.need_op {
             let mut output_vec = input_col_change.to_parts().0.to_vec();
             let len = input_col_change.to_parts().1;
             output_vec.push(Some(len));
-            let out_col_change = ColIndexMapping::new(output_vec, len + 1);
-            (changelog, out_col_change)
+            ColIndexMapping::new(output_vec, len + 1)
+        } else {
+            input_col_change
+        };
+
+        let mut output_vec = out_col_change.to_parts().0.to_vec();
+        let len = out_col_change.to_parts().1;
+        let out_col_change = if self.base.stream_key().is_none() {
```

Comment:
I think we can use `need_changelog_row_id` instead of `stream_key()` to make this logic more clear.